kit/popups/src/Menu/UIMenuContainer.tsx

```diff
@@ -0,0 +1,189 @@
+        bottom: windowDimensions.height - UILayoutConstant.contentOffset - menuSize.height,
+        left: UILayoutConstant.contentOffset,
+    };
+}
+
+function useMenuLocation(
+    targetDimensions: Dimensions | null,
+    windowDimensions: ScaledSize,
+    menuSize: Size,
+): Location | null {
+    return React.useMemo(() => {
+        if (!menuSize.height || !menuSize.width || !targetDimensions) {
+            return null;
+        }
+
+        let top: number | undefined = targetDimensions.y + targetDimensions.height;
+        let bottom: number | undefined;
+        let left: number = targetDimensions.x;
+
+        const boundaries = getBoundaries(windowDimensions, menuSize);
+
+        if (left > boundaries.right) {
+            left = boundaries.right;
+        }
+        if (left < boundaries.left) {
+            left = boundaries.left;
+        }
+
+        if (top > boundaries.bottom) {
+            top = undefined;
+            bottom = windowDimensions.height - targetDimensions.y;
+        }
+
+        return {
+            left,
+            top,
+            bottom,
+        };
+    }, [targetDimensions, windowDimensions, menuSize]);
+}
+
+export function UIMenuContainer({ children, visible, targetRef, onClose }: UIMenuContainerProps) {
+    const theme = useTheme();
+    const windowDimensions = useWindowDimensions();
+    const targetDimensions = useTargetDimensions(visible, targetRef, windowDimensions);
+
+    const [menuSize, setMenuSize] = React.useState<Size>(initialSize);
+    const onLayout = React.useCallback(
+        function onLayout(event: LayoutChangeEvent) {
+            const width = Math.round(event.nativeEvent.layout.width);
+            const height = Math.round(event.nativeEvent.layout.height);
+            if (menuSize.width !== width || menuSize.height !== height) {
+                setMenuSize({ width, height });
+            }
+        },
+        [menuSize],
+    );
+    const menuLocation = useMenuLocation(targetDimensions, windowDimensions, menuSize);
+
+    const onPressIn = React.useCallback(
+        function onPressIn() {
+            onClose();
+        },
+        [onClose],
+    );
+
+    const { color: shadowColor, opacity: shadowOpacity } = useColorParts(
+        ColorVariants.BackgroundOverlay,
+    );
+    const styles = useStyles(theme, menuLocation, shadowColor, shadowOpacity);
+
+    if (!visible) {
+        return null;
+    }
+
+    return (
+        <Portal absoluteFill>
+            <TouchableOpacity onPressIn={onPressIn} style={StyleSheet.absoluteFill} />
+            <Animated.View
+                style={styles.container}
+                entering={FadeIn.duration(UIConstant.menu.animationTime)}
+                exiting={FadeOut.duration(UIConstant.menu.animationTime)}
+                onLayout={onLayout}
+            >
+                <ShadowView style={styles.shadowContainer}>{children}</ShadowView>
```

Comment:
Have you tried to use `Animated.createAnimatedComponent` to remove wrapping `Animated.View`?

Reply:
I tried it. There were problems with `ShadowView` performance, especially on android. It started lagging.

Reply:
Ok :)